client/state/shortcodes/reducer.js

```diff
@@ -50,6 +52,32 @@ export const requesting = createReducer( {}, {
  * @return {Object}        Updated state
  */
 export const items = createReducer( {}, {
+	[ MEDIA_ITEMS_RECEIVE ]: ( state, { siteId, data } ) => {
+		if ( ! state.hasOwnProperty( siteId ) ) {
+			return state;
+		}
+
+		let media;
+		if ( Array.isArray( data.media ) ) {
+			media = data.media;
+		} else {
+			media = [ data ];
+		}
+
+		const updatedIds = media.map( ( item ) => item.ID.toString() );
+
+		return Object.assign( {}, state, {
+			[ siteId ]: pickBy( state[ siteId ], ( shortcode ) => {
+				const parsed = Shortcode.parse( shortcode.shortcode );
+				if ( parsed.tag !== 'gallery' || ! parsed.attrs || ! parsed.attrs.named || ! parsed.attrs.named.ids ) {
```

Comment:
Thinking `_.has` could reduce the verbosity of the nested property check, and arguably more human-readable:

```js
if ( parsed.tag !== 'gallery' || ! has( parsed, 'attrs.named.ids' ) {
```